### PR TITLE
avoid EncodingWarning in bokeh.resources and bokeh.themes.theme 

### LINKED ISSUE #12711: [BUG] EncodingWarning in bokeh.resources and bokeh.themes.Theme(...)
### Software versions

Python version      :  3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:35:26) [GCC 10.4.0]
IPython version     :  8.6.0
Tornado version     :  6.2
Bokeh version       :  2.4.3
BokehJS static path :  /home/graingert/anaconda3/envs/dask-distributed/lib/python3.10/site-packages/bokeh/server/static
node.js version     :  (not installed)
npm version         :  (not installed)
Operating system    :  Linux-5.15.0-56-generic-x86_64-with-glibc2.35

### Browser name and version

_No response_

### Jupyter notebook / Jupyter Lab version

_No response_

### Expected behavior

no EncodingWarnings emitted

### Observed behavior

EncodingWarnings emitted

### Example code

```Python
python -X warn_default_encoding -c "from bokeh.plotting import figure; from bokeh.themes import Theme; Theme('demo.yaml')"
/home/graingert/anaconda3/envs/dask-distributed/lib/python3.10/site-packages/bokeh/resources.py:128: EncodingWarning: 'encoding' argument not specified
  with open(join(ROOT_DIR, "_sri.json")) as f:
/home/graingert/anaconda3/envs/dask-distributed/lib/python3.10/site-packages/bokeh/themes/theme.py:165: EncodingWarning: 'encoding' argument not specified
  with open(filename) as f:
```

### Stack traceback or browser console output

_No response_

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by graingert**
- [x] issues: fixes #12711
- [x] tests added / passed
- [ ] release document entry (if new feature or API change)

### CODE REVIEW

File: src/bokeh/resources.py [hunk 1]
```diff
@@ -127,7 +127,7 @@ def get_all_sri_hashes() -> dict[str, Hashes]:
     global _SRI_HASHES
 
     if not _SRI_HASHES:
-        with open(join(ROOT_DIR, "_sri.json")) as f:
+        with open(join(ROOT_DIR, "_sri.json", "rb")) as f:
```
Comment (graingert) [inline src/bokeh/resources.py]:
```suggestion
        with open(join(ROOT_DIR, "_sri.json"), "rb") as f:
```

**Comment by bryevdv** [comment]
I am not really sure this is necessary or advised. The files being opened will only ever contain ASCII. Per pep 597:

> If ASCII is assumed, this isn’t a bug

cc @bokeh/core

**Comment by bryevdv** [comment]
Well, the SRI files certainly will. I guess an end user could create a theme file with whatever they want, but a proper solution for that is to allow users to specify their own encoding with the filename.

**Comment by graingert** [comment]
yaml is defined as supporting only UTF-8, UTF-16 and UTF-32 encodings. Opening the yaml file in bytes mode allows the yaml processor to read the BOM to determine the encoding

**Comment by bryevdv** [comment]
> Opening the yaml file in bytes mode allows the yaml processor to read the BOM to determine the encoding

@graingert Do you have a reference for this? The docs at https://pyyaml.org/wiki/PyYAMLDocumentation merely state:

> A byte string or a file must be encoded with utf-8, utf-16-be or utf-16-le encoding. yaml.load detects the encoding by checking the BOM (byte order mark) sequence at the beginning of the string/file. If no BOM is present, the utf-8 encoding is assumed.

and do not mention (or demonstrate) opening in binary mode.

**Comment by graingert** [comment]
yeah that's the reference I had in mind:

when you open a yaml file with `rb` yaml can auto-detect the encoding:
```
>>> with open("example.yaml", "w", encoding="utf-16") as f:
...     yaml.dump({}, f)
... 
>>> with open("example.yaml", "rb") as f:
...     yaml.safe_load(f)
... 
{}
```

but if you use the "locale" encoding the determine_encoding feature doesn't work:
```
>>> with open("example.yaml") as f:
...     yaml.safe_load(f)
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/home/graingert/.virtualenvs/testing311/lib/python3.11/site-packages/yaml/__init__.py", line 125, in safe_load
    return load(stream, SafeLoader)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/graingert/.virtualenvs/testing311/lib/python3.11/site-packages/yaml/__init__.py", line 79, in load
    loader = Loader(stream)
             ^^^^^^^^^^^^^^
  File "/home/graingert/.virtualenvs/testing311/lib/python3.11/site-packages/yaml/loader.py", line 34, in __init__
    Reader.__init__(self, stream)
  File "/home/graingert/.virtualenvs/testing311/lib/python3.11/site-packages/yaml/reader.py", line 85, in __init__
    self.determine_encoding()
  File "/home/graingert/.virtualenvs/testing311/lib/python3.11/site-packages/yaml/reader.py", line 124, in determine_encoding
    self.update_raw()
  File "/home/graingert/.virtualenvs/testing311/lib/python3.11/site-packages/yaml/reader.py", line 178, in update_raw
    data = self.stream.read(size)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen codecs>", line 322, in decode
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte
```

**Comment by graingert** [comment]
regarding the SRI files omitting the encoding kwarg may be a false positive of PEP 597 it still throws the warning on import for downstream projects.

I could have passed `encoding="ascii"` here, but opening the the json file in bytes mode also supports auto-detecting the encoding so I opted for that

**Comment by bryevdv** [comment]
@graingert sorry I got distracted away from this. I think I'm OK with putting this in but I would feel better about it if there were some tests to go along, that could serve to document the expectations in different situations. Would you be able to contribute a few unit tests for this PR?

**Comment by graingert** [comment]
@bryevdv I added a test for auto-detecting encoding of yaml theme files, but I've not added a test for get_all_sri_hashes because it would involve monkeypatching the file for `src/bokeh/_sri.json` and I don't think it's worth it. The other way to test it would be to configure pytest `filterwarnings = error` and set `env PYTHONWARNDEFAULTENCODING=1`, but that would be quite a large endeavour

**Comment by bryevdv** [comment]
@graingert I was prepared to merge before [4d78186](https://github.com/bokeh/bokeh/pull/12712/commits/4d781869c1ab7833161ec8af97f2049a91f1fe1a) I think those changes would require more scrutiny. Do you want to revert that commit and we can merge this and consider further changes in a separate PR?

**Comment by graingert** [comment]
@bryevdv woops that was my plan but I pushed to the wrong branch

**Review by bryevdv — APPROVED** [review]

**Comment by bryevdv** [comment]
Thanks @graingert !

**Pull request merged**